File: src/trading/indicators.ts
/* ─── Indicateurs techniques (calculés sur clôtures journalières) ───────── */

/** Moyenne mobile simple. Renvoie null tant que la fenêtre n'est pas pleine. */
export function sma(values: number[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(values.length).fill(null);
  let sum = 0;
  for (let i = 0; i < values.length; i++) {
    sum += values[i];
    if (i >= period) sum -= values[i - period];
    if (i >= period - 1) out[i] = sum / period;
  }
  return out;
}

/** RSI de Wilder. */
export function rsi(values: number[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(values.length).fill(null);
  if (values.length <= period) return out;

  let avgGain = 0;
  let avgLoss = 0;
  for (let i = 1; i <= period; i++) {
    const diff = values[i] - values[i - 1];
    if (diff > 0) avgGain += diff;
    else avgLoss -= diff;
  }
  avgGain /= period;
  avgLoss /= period;
  out[period] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);

  for (let i = period + 1; i < values.length; i++) {
    const diff = values[i] - values[i - 1];
    const gain = diff > 0 ? diff : 0;
    const loss = diff < 0 ? -diff : 0;
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
    out[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  }
  return out;
}

/** ATR de Wilder (amplitude moyenne vraie) — sert à placer le stop. */
export function atr(
  highs: number[],
  lows: number[],
  closes: number[],
  period: number
): (number | null)[] {
  const n = closes.length;
  const out: (number | null)[] = new Array(n).fill(null);
  if (n <= period) return out;

  const tr: number[] = new Array(n).fill(0);
  tr[0] = highs[0] - lows[0];
  for (let i = 1; i < n; i++) {
    tr[i] = Math.max(
      highs[i] - lows[i],
      Math.abs(highs[i] - closes[i - 1]),
      Math.abs(lows[i] - closes[i - 1])
    );
  }

  let avg = 0;
  for (let i = 1; i <= period; i++) avg += tr[i];
  avg /= period;
  out[period] = avg;

  for (let i = period + 1; i < n; i++) {
    avg = (avg * (period - 1) + tr[i]) / period;
    out[i] = avg;
  }
  return out;
}

File: backend/app/indicators.py
"""Indicateurs techniques — logique pure (aucune dépendance tierce).

Réplique côté backend les indicateurs du frontend (cohérence des signaux)
et en ajoute (volume moyen) pour le scoring.
"""

from __future__ import annotations

import math
from dataclasses import dataclass


@dataclass
class Candle:
    time: int
    open: float
    high: float
    low: float
    close: float
    volume: float = 0.0


def sma(values: list[float], period: int) -> list[float | None]:
    out: list[float | None] = [None] * len(values)
    total = 0.0
    for i, v in enumerate(values):
        total += v
        if i >= period:
            total -= values[i - period]
        if i >= period - 1:
            out[i] = total / period
    return out


def rsi(values: list[float], period: int) -> list[float | None]:
    """RSI de Wilder."""
    out: list[float | None] = [None] * len(values)
    if len(values) <= period:
        return out

    gain = loss = 0.0
    for i in range(1, period + 1):
        diff = values[i] - values[i - 1]
        if diff > 0:
            gain += diff
        else:
            loss -= diff
    gain /= period
    loss /= period
    out[period] = 100.0 if loss == 0 else 100.0 - 100.0 / (1 + gain / loss)

    for i in range(period + 1, len(values)):
        diff = values[i] - values[i - 1]
        g = diff if diff > 0 else 0.0
        ll = -diff if diff < 0 else 0.0
        gain = (gain * (period - 1) + g) / period
        loss = (loss * (period - 1) + ll) / period
        out[i] = 100.0 if loss == 0 else 100.0 - 100.0 / (1 + gain / loss)
    return out


def atr(candles: list[Candle], period: int) -> list[float | None]:
    """ATR de Wilder (amplitude moyenne vraie)."""
    n = len(candles)
    out: list[float | None] = [None] * n
    if n <= period:
        return out

    tr = [0.0] * n
    tr[0] = candles[0].high - candles[0].low
    for i in range(1, n):
        c = candles[i]
        prev = candles[i - 1].close
        tr[i] = max(c.high - c.low, abs(c.high - prev), abs(c.low - prev))

    avg = sum(tr[1 : period + 1]) / period
    out[period] = avg
    for i in range(period + 1, n):
        avg = (avg * (period - 1) + tr[i]) / period
        out[i] = avg
    return out


def avg_volume(candles: list[Candle], period: int) -> float | None:
    """Volume moyen sur les `period` dernières bougies (hors dernière)."""
    if len(candles) < period + 1:
        return None
    window = candles[-period - 1 : -1]
    vols = [c.volume for c in window if c.volume > 0]
    if not vols:
        return None
    return sum(vols) / len(vols)


def stdev(values: list[float]) -> float:
    n = len(values)
    if n < 2:
        return 0.0
    mean = sum(values) / n
    return math.sqrt(sum((v - mean) ** 2 for v in values) / (n - 1))
